File: psr/lib/psr.py
prods1,prods2 = [1],[1]
 
for p in [2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73]:
    prods1 += [x*p for x in prods1]
 
for p in [79,83,89,97,101,103,107,109,113,127,131,137,139,149,151,157,163,167,173,179,181]:
    prods2 += [x*p for x in prods2]

#print prods1

#print prods2
 
prods1.sort()
prods2.sort(reverse=True)
target = (prods1[-1]*prods2[0])**0.5
l1,l2 = len(prods1),len(prods2)

print target
#print l2
 
i=j=best=0
while i<l1 and j<l2:
    current = prods1[i]*prods2[j]
    if current>target:
        j+=1
    elif current<target:
        if current>best: best=current
        i+=1
 
print best%(10**16)

File: psr/lib/psr.rb
require 'primes'

class PSR
  def self.for(n = 190)
    ps = Primes.for(n)
    fh =[1]
    hop = (ps.length/2).to_i
    ps[0..hop].each do |prime|
      fh += fh.map {|value| value * prime }
    end
    sh =[1]
    ps[hop + 1..ps.length].each do |prime|
      sh += sh.map {|value| value * prime }
    end

    fh = fh.sort
    sh = sh.sort.reverse

    sqrt_i = (fh[-1] * sh[0])**0.5
    l1, l2 = fh.length, sh.length

    i = j = psr = 0
    while (i < l1 and j < l2)
      current = fh[i] * sh[j]
      if current > sqrt_i
        j += 1
      elsif current < sqrt_i
         psr = current if current > psr
         i += 1
      end
    end
    return psr
  end

  def self.answer_for(n = 190)
    return self.for(n) % (10**16)
  end
end